src/java/org/apache/cassandra/index/sai/disk/v1/sortedterms/TrieRangeIterator.java

```diff
@@ -0,0 +1,268 @@
+
+        if (start != null)
+            initializeWithLeftBound(root, start.asComparableBytes(BYTE_COMPARABLE_VERSION), admitPrefix, limit != null);
+        else
+            initializeNoLeftBound(root, limit != null ? limit.next() : 256);
+    }
+
+    public Iterator<Pair<ByteSource, Long>> iterator()
+    {
+        return new AbstractIterator<Pair<ByteSource, Long>>()
+        {
+            @Override
+            protected Pair<ByteSource, Long> computeNext()
+            {
+                final long node = nextPayloadedNode();
+                if (node == -1 || getCurrentPayload() == -1)
+                {
+                    return endOfData();
+                }
+                ByteSource byteSource = collector.toByteComparable().asComparableBytes(ByteComparable.Version.OSS42);
+                return Pair.create(byteSource, getCurrentPayload());
+            }
+        };
+    }
+
+    public long getCurrentPayload()
+    {
+        return getPayload(buf, payloadPosition(), payloadFlags());
+    }
+
+    private long getPayload(ByteBuffer contents, int payloadPos, int bytes)
+    {
+        if (bytes == 0)
+        {
+            return TrieTermsDictionaryReader.NOT_FOUND;
+        }
+        return SizedInts.read(contents, payloadPos, bytes);
+    }
+
+    private long nextPayloadedNode()
+    {
+        if (first)
+        {
+            first = false;
+            if (stack == null) return -1;
+            return stack.node;
+        }
+        long toReturn = next;
+        if (next != -1)
+            next = advanceNode();
+        return toReturn;
+    }
+
+    private void initializeWithLeftBound(long root, ByteSource startStream, boolean admitPrefix, boolean atLimit)
+    {
+        IterationPosition prev = null;
+        int childIndex;
+        int limitByte;
+        long payloadedNode = -1;
+
+        try
+        {
+            // Follow start position while we still have a prefix, stacking path and saving prefixes.
+            go(root);
+            while (true)
+            {
+                int s = startStream.next();
+                childIndex = search(s);
+
+                // For a separator trie the latest payload met along the prefix is a potential match for start
+                if (admitPrefix)
+                    if (childIndex == 0 || childIndex == -1)
+                    {
+                        if (payloadFlags() != 0)
+                            payloadedNode = position;
+                    }
+                    else
+                        payloadedNode = -1;
+
+                limitByte = 256;
+                if (atLimit)
+                {
+                    limitByte = limit.next();
+                    if (s < limitByte)
+                        atLimit = false;
+                }
+                if (childIndex < 0)
+                    break;
+
+                int transitionByte = transitionByte(childIndex);
+                collector.add(transitionByte);
```

Comment:
If it weren't for these two lines, I want to say this `initializeWithLeftBound()` would be identical to the `initializeWithLeftBound()` in `ValueIterator`. Maybe it could be a pluggable behavior and a no-op for `ValueIterator`.

Reply:
I was a bit confused by this and then realised that `ValueIterator` was deleted following a comment by @adelapena that it was not actually used anywhere. If `ValueIterator` is reinstated at any point then I agree that we should optimize them properly.

Reply:
EDIT AGAIN: The above change did not work as expected so I have pulled the trie operations from the `SortedTermsReader` entirely and have added a new class to handle value lookups in the trie.

The `TrieTermsDictionaryReader.exactMatch` did not work because, as the name says, it does an exact match and the `PrimaryKeyMap` can lookup partial keys, as in the `DecoratedKey` without the associated `Clustering`. This is effectively a prefix lookup into the trie and the trie is expected (by the `PrimaryKeyMap`) to return the first key that matches the prefix.

I have a created a `TriePrefixSearcher` class to perform this operation. This currently only performs to the specific requirements of the `PrimaryKeyMap` it can easily be extended to be generic and handle deep prefix lookups that return all the payloads under the prefix.